src/crypto/chacha_poly_aead.cpp

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <crypto/chacha_poly_aead.h>
+
+#include <crypto/common.h>
+#include <crypto/poly1305.h>
+#include <support/cleanse.h>
+
+#include <assert.h>
+#include <string.h>
+
+#include <cstdio>
+#include <limits>
+
+#ifndef HAVE_TIMINGSAFE_BCMP
+
+int timingsafe_bcmp(const unsigned char* b1, const unsigned char* b2, size_t n)
+{
+    const unsigned char *p1 = b1, *p2 = b2;
+    int ret = 0;
+
+    for (; n > 0; n--)
+        ret |= *p1++ ^ *p2++;
+    return (ret != 0);
+}
+
+#endif // TIMINGSAFE_BCMP
+
+ChaCha20Poly1305AEAD::ChaCha20Poly1305AEAD(const unsigned char* K_1, size_t K_1_len, const unsigned char* K_2, size_t K_2_len)
+{
+    assert(K_1_len == CHACHA20_POLY1305_AEAD_KEY_LEN);
+    assert(K_2_len == CHACHA20_POLY1305_AEAD_KEY_LEN);
+    m_chacha_main.SetKey(K_1, CHACHA20_POLY1305_AEAD_KEY_LEN);
+    m_chacha_header.SetKey(K_2, CHACHA20_POLY1305_AEAD_KEY_LEN);
+
+    // set the cached sequence number to uint64 max which hints for an unset cache.
+    // we can't hit uint64 max since the rekey rule (which resets the sequence number) is 1GB
+    m_cached_aad_seqnr = std::numeric_limits<uint64_t>::max();
+}
+
+bool ChaCha20Poly1305AEAD::Crypt(uint64_t seqnr_payload, uint64_t seqnr_aad, int aad_pos, unsigned char* dest, size_t dest_len /* length of the output buffer for sanity checks */, const unsigned char* src, size_t src_len, bool is_encrypt)
+{
+    // check buffer boundaries
+    if (
+        // if we encrypt, make sure the source contains at least the expected AAD and the destination has at least space for the source + MAC
+        (is_encrypt && (src_len < CHACHA20_POLY1305_AEAD_AAD_LEN || dest_len < src_len + POLY1305_TAGLEN)) ||
+        // if we decrypt, make sure the source contains at least the expected AAD+MAC and the destination has at least space for the source - MAC
+        (!is_encrypt && (src_len < CHACHA20_POLY1305_AEAD_AAD_LEN + POLY1305_TAGLEN || dest_len < src_len - POLY1305_TAGLEN))) {
+        return false;
+    }
+
+    unsigned char expected_tag[POLY1305_TAGLEN], poly_key[POLY1305_KEYLEN];
+    memset(poly_key, 0, sizeof(poly_key));
+    m_chacha_main.SetIV(htole64(seqnr_payload));
```

Comment:
are you sure the `htole64` is correct here? `SetIV` is implemented as
```c++
void ChaCha20::SetIV(uint64_t iv)
{
    input[14] = iv;
    input[15] = iv >> 32;
}
```
where `input` is an array of 32-bit unsigned integers that are (afaik) always used in LE fashion with `ReadLE32` / `WriteLE32` when doing byte operations.
`htole64` flips the bytes of an integer on a big-endian platform
so on LE, the output looks like
```
0      1       2        3        4        5        6        7
iv0..7 iv8..15 iv16..23 iv24..31 iv32..39 iv40..47 iv48..55 iv56.63
```
on BE, however you'd get
```
0        1        2       3      4       5         6        7
iv24..31 iv16..23 iv8..15 iv0..7 iv56.63 iv48..55  iv40..47 iv32..39
```
I might be wrong, but this needs testing on a big-endian platform.

In any case as you always use `SetIV` with `htole64`, might make sense to move it into it.

Reply:
I think it is necessary, because even though `input` is always used with `ReadLE32`/`WriteLE32` when you shift the iv and split it into `input[14]` and `input[15]` you'll get different results in BE or LE platforms.
(unless you use `htole64` or something equivalent to make sure that the bits are sorted the same way)

But I do agree that this should be part of the chacha20 implementation. because this is what requires that everything will be LE.

Reply:
There should not be any need for byteswapping. The IV is specified as an integer _before serialization_. The ChaCha20 implementation uses the LE serialization of that number, even on BE platforms.

Reply:
> I think it is necessary, because even though input is always used with 

If this is true, I still think it's done in the wrong way if the goal is to be consistent between LE and BE, you'd need a different kind of word-level swap. But I think @sipa is right here.

Reply:
The only thing that bothers me is the split using r/l shift.
Chacha20 defines everything as LE serialized/deserialized and if you shift than you bypass this requirement on BE platforms.

(unless `seqnr_payload` is somehow already LE and not the host native endianess)

Reply:
seqnr_payload is an integer. The IV argument to ChaCha20 is an integer. Byte orders are completely irrelevant. This is a pointless discussion.

Reply:
Ok, I went back to look at the definitions and tested, and right/left shifts work the same in all platforms (which actually is the sane thing, otherwise it will be a bit UB). you and Wladimir are right, the `htole64` isn't needed here.

Reply:
Thanks for looking into this...
I'm actually unsure also by comparing against openSSH's implementation:
* https://github.com/openssh/openssh-portable/blob/master/chacha.c#L85
* https://github.com/openssh/openssh-portable/blob/master/cipher-chachapoly.c#L65

Going to setup a BE machine to test against various test vectors.

Reply:
> and right/left shifts work the same in all platforms 

C++ has plenty of cruel and unusual cases of UB and IDB, but this luckily isn't one

> I'm actually unsure also by comparing against openSSH's implementation:

Looking at these, they have `POKE_U64` which pokes a unsigned 64-bit integer into memory (big-endian byte order, see https://github.com/openssh/openssh-portable/blob/master/sshbuf.h#L266). This is used to poke the IV to a byte buffer `iv`. e.g.
```c++
iv[0] = (seqnr >> 56) & 0xff;
iv[1] = (seqnr >> 48) & 0xff;
iv[2] = (seqnr >> 40) & 0xff;
iv[3] = (seqnr >> 32) & 0xff;
iv[4] = (seqnr >> 24) & 0xff;
iv[5] = (seqnr >> 16) & 0xff;
iv[6] = (seqnr >> 8) & 0xff;
iv[7] = seqnr & 0xff;
```
Then, in `chacha_ivsetup` (`U8TO32_LITTLE` is defined here: https://github.com/openssh/openssh-portable/blob/master/chacha.c#L27)
```c++
x->input[14] = U8TO32_LITTLE(iv + 0);
x->input[15] = U8TO32_LITTLE(iv + 4);
```
So if I get this right, this is equivalent to:
```
x->input[14] = (((seqnr >> 56) & 0xff)      ) |
               (((seqnr >> 48) & 0xff) <<  8) |
               (((seqnr >> 40) & 0xff) << 16) |
               (((seqnr >> 32) & 0xff) << 24);

x->input[15] = (((seqnr >> 24) & 0xff)      ) |
               (((seqnr >> 16) & 0xff) <<  8) |
               (((seqnr >>  8) & 0xff) << 16) |
               (( seqnr        & 0xff) << 24);
```
Which is weird in some sense, but not dependent on platform.

Reply:
Thanks for clearing up.
@sipa and @laanwj are def. correct. 
Removed all byteswappings.

Reply:
The only important part is, that "producers" or the AEAD stream (which is not part of this PR, we only intend to test against already produced test vectors) need to make sure that the 3 byte length is correctly in LE order (which is easy to screw up because htole32 won't work in the 24bit case).